### ISSUE TITLE
For `ink_e2e.v`: Add instance for Coq to find in `subxt.v`

### ISSUE THREAD

**Issue by InfiniteEchoes** [formal-land/coq-of-rust#195]
Simpler version of the bug has been abstracted as a code snippet that can be added right at the end of `examples/traits/traits_params.v`:

```coq
Require Import CoqOfRust._std.fmt.

Module Module1.
  Unset Primitive Projections.
  Module Struct1.
    Record t (T1 T2 : Set)
      `{Debug.Trait T1}
      `{SomeTrait.Trait T2}
    : Set := { }.
  End Struct1.
  Global Set Primitive Projections.
  Definition Struct1 := Struct1.t.
End Module1.

Module Module2.
  Unset Primitive Projections.
  Module Struct2.
    Record t : Set := { }.
  End Struct2.
  Global Set Primitive Projections.
  Definition Struct2 := Struct2.t.
End Module2.

Definition TypeDef1 (T : Set)  : Set :=  Module1.Struct1 T Module2.Struct2. (* <-- TODO: Make this line compile *)
```

The complicated and original version stems from one of the dependencies for `ink_e2e.v`, the crate `subxt`.

```coq
(* 
pub trait Config: 'static {
    type Index: Debug + Copy + DeserializeOwned + Into<u64>;
    type Hash: Debug + Copy + Send + Sync + Decode + AsRef<[u8]> + Serialize + DeserializeOwned + Encode + PartialEq;
    type AccountId: Debug + Clone + Serialize;
    type Address: Debug + Encode + From<Self::AccountId>;
    type Signature: Debug + Encode;
    type Hasher: Debug + Hasher<Output = Self::Hash>;
    type Header: Debug + Header<Hasher = Self::Hasher> + Send + DeserializeOwned;
    type ExtrinsicParams: ExtrinsicParams<Self::Index, Self::Hash>;
}
*)
Module config.
(* 1: `SomeTrait` in the `trait_params` code *)
  Module Config.
    Class Trait (Self : Set)
                {Index
                Hash
                AccountId
                Address
                Signature
                Hasher
                Header
                ExtrinsicParams
                : Set} 
    : Set := {
    (* NOTE: we make a very rough translation here... *)
        Index := Index;
        Hash := Hash;
        AccountId := AccountId;
        Address := Address;
        Signature := Signature;
        Hasher := Hasher;
        Header := Header;
        ExtrinsicParams := ExtrinsicParams;
    }.
End config.

Module extrinsic_params.
    (* 2: Struct1 in the example code *)
    (* pub struct BaseExtrinsicParams<T: Config, Tip: Debug> { /* private fields */ } *)
    Unset Primitive Projections.
    Module BaseExtrinsicParams.
      Record t (T Tip : Set) 
        `{Config.Trait T}
        `{Debug.Trait Tip}
      : Set := { }.
    End BaseExtrinsicParams.
    Global Set Primitive Projections.
    Definition BaseExtrinsicParams := BaseExtrinsicParams.t.
  End extrinsic_params. 

Module polkadot.
  (*3 : Struct2 in the example code*)
  (* pub struct PlainTip { /* private fields */ } *)
  Unset Primitive Projections.
  Module PlainTip.
    Record t : Set := { }.
  End PlainTip.
  Global Set Primitive Projections.
  Definition PlainTip := PlainTip.t.

  (* 4: The Definition in the code snippet *)
  (* pub type PolkadotExtrinsicParams<T> = BaseExtrinsicParams<T, PlainTip>; *)
  Definition PolkadotExtrinsicParams (T : Set) : Set := extrinsic_params.BaseExtrinsicParams T PlainTip. (* <-- TODO: Make this line compile *)
End polkadot.
```

**Comment by InfiniteEchoes** [comment]
Below is the correct way to add the missing instances for example code:

```coq
Global Instance I : SomeTrait.Trait Module2.Struct2. Admitted.
Global Instance I1 {T : Set} : Debug.Trait T. Admitted.

Definition TypeDef1 (T : Set) : Set := Module1.Struct1 T Module2.Struct2 _ _ _ _ _ _.
```

Now I'll work on working it back to `subxt` an fix the bug.

**Comment by InfiniteEchoes** [comment]
On the subxt side it turns out that `config` trait is slightly more compilcatd as it gets a lot of associated types within. Bartex says that he will work on changing how we translate those typedefs and I'll wait for his updates. Note that in current situation, even if I have given out the correct instance with those typedefs as the parameters for these instances, Coq will report error as it cannot find the correct instances for just these typedefs.

In conclusion,  I think this issue can be closed.
